(WIP) Add a add completeness test for all types with CacheInfo method

kevinkengne opened the pull request:
Five types seems to have a CacheInfo() method. The cache Key for one of them (`DatacentersRequest`) is a constant. Therefore, I didn't add a test for that type. Please let me know what can be improved on this PR.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=9480)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=9480)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=9480).</sub>

kevinkengne commented:
resolves #9437

dnephin reviewed (commented):
Thank you for working on these tests! This looks like an excellent start.

One detail I missed in the issue is about how the key is used. The reason that the `Token` field is not represented in the key is because in the cache implementation we prepend the key with `Datacenter` and `Token` https://github.com/hashicorp/consul/blob/57701695c3b549b719ecd192a8f9d8bd2730341b/agent/cache/cache.go#L395 So we expect that Token and Datacenter fields will not be represented in the `Key` field.

I think these tests make good use of gofuzz, but I don't think they will catch the original problem, because they are only testing specific fields. What I imagined was something like this (this is rough code, may not actually compile):

```go
req := NodeSpecificRequest{}
fuzzer.Fuzz(&req)
origKey := req.CacheInfo().Key
reqValue := reflect.ValueOf(req)
for i := 0; i < reqValue.NumField(); i++ {
    field := reqValue.Field(i)
    // TODO: skip Datacenter, Token, and other ignored fields

    fuzzer.Fuzz(field.Addr().Interface())
    key := req.CacheInfo().Key
    if key == origKey {
        t.Fatal("expected field %v to be represented in the CacheInfo.Key", reqValue.Type().Field(i).Name)
    }
}
```

I believe a test like this will catch the problem that occurs when a new field is added to the request struct, but not to the cache key.

kevinkengne commented:
@dnephin, thanks for the detailed explanation. That makes sense! I'm working towards implementing your suggestions. I just have one issue where certain fields e.g `NodeMetaFilters` on `ServiceSpecificRequest` aren't addressable, (that field is of type `map[string]string` and from what I gather in the godoc of the reflect package and from the panics I get in the tests, only an element of a slice, an element of an addressable array, a field of an addressable struct, or the result of dereferencing a pointer are addressable. How can I get to the underlying pointer for these unaddressable fields in order to be able to call `fuzzer.Fuzz()` on them a second time?

dnephin commented:
Ah good question! I guess that's the `field.Addr().Interface()` line? I thought all of those would be fields of an addressable struct.

I found https://stackoverflow.com/questions/6395076/using-reflect-how-do-you-set-the-value-of-a-struct-field/6396678#6396678, which suggests the problem may be with `reqValue := reflect.ValueOf(req)`.

I think to allow it to be addressable that line needs to be changed to `reqValue := reflect.ValueOf(&req).Elem()` (the value of the pointer to the struct, then using `Elem()` to get the value of the struct with is addressable).

kevinkengne commented:
@dnephin Thanks, I've implemented the changes. Currently the fields ignored by the tests are hard coded as I couldn't think of a dynamic way of getting them for each type. However, the tests will be able to catch the problem that occurs when a new field is added to the request struct, but not to the cache key.

Two things to note with the current state of the PR:

- `EnterpriseMeta` is part of the cache key, however as it is an anonymous field I believe it might be confusing the fuzzer so I've added it to ignored fields.
-  `ServiceDumpRequest.ServiceKind` is only used as part of the cache key if `ServiceDumpRequest.UseServiceKind` is true. Therefore, I've also added it to the ignored fields for that type.

dnephin commented:
@kevinkengne thank you for working on this! This looks great!

I tried it out locally and ran into some test flakes. I think what was happening was that the `fuzzer.Fuzz()` happened to generate the same data again, which caused the test to fail. To handle this I re-fuzz any time the value has not changed. I think that's what you were doing with the bool already, I just extended it a bit to all types.

I also made a small change for handling `EnterpriseMeta` so that we can test those fields in the enterprise fork. My change there caused a conflict, so I'm going to rebase this PR on master.

If the tests are happy after that change I will merge.

dnephin commented:
I found a few more cache types in other files, so I started to add tests for those in #9556. I found some surprising omissions from the cache keys, which may be bugs. I'll have to look into that more.

dnephin reviewed (approved):
LGTM!

kevinkengne commented:
Thanks!
